h2/src/main/org/h2/command/dml/Select.java

```diff
@@ -151,8 +168,51 @@ public void setGroupBy(ArrayList<Expression> group) {
         return group;
     }
 
-    public HashMap<Expression, Object> getCurrentGroup() {
-        return currentGroup;
+    /**
+     * Is there currently a group-by active
+     */
+    public boolean isCurrentGroup() {
+        return currentGroupByExprData != null;
+    }
+
+    /**
+     * Get the group-by data for the current group and the passed in expression.
+     */
+    public Object getCurrentGroupExprData(Expression expr) {
+        Integer index = exprToIndexInGroupByData.get(expr);
+        if (index == null) {
+            return null;
+        }
+        cachedLookup = expr;
+        cachedLookupIndex = index;
+        return currentGroupByExprData[index];
+    }
+
+    /**
+     * Set the group-by data for the current group and the passed in expression.
+     */
+    public void setCurrentGroupExprData(Expression expr, Object obj) {
+        if (cachedLookup == expr) {
+            assert currentGroupByExprData[cachedLookupIndex] == null;
+            currentGroupByExprData[cachedLookupIndex] = obj;
```

Comment:
We should return from the method here, but `exprToIndexInGroupByData` can be empty at this point.

If `exprToIndexInGroupByData` have such mapping we set `currentGroupByExprData[...]` again few lines below, that is not needed. But if `exprToIndexInGroupByData` is empty the mapping is inserted below.

Tests (with proper assertions) are fine, but logic seems to be weird enough. From my point of view we should not have empty map if cached values are set.

Also this optimization did not improve execution times on my tests.

May be it is not needed so much?

Also I think that this optimization can be done in another way. Some interface for allocation of `AggregateData` can be introduced. `Aggregate`, `JavaAggregate`, and `ExpressionColumn` can implement this interface. And instead of this setter method an another getOrSet method can lookup a value or allocate, set, and return a new value using this interface method from a caller. This way should be simpler.

Reply:
I didn't notice before that this optimization sets value more than one time.

Reply:
Yeah the variability in this test is a pain